pkg/operator/operators/clustermanager/storageversionmigration/precheck.go

```diff
@@ -0,0 +1,125 @@
+}
+
+// Purpose: This function is used to solve a MCE 2.4 upgrade to 2.5 cases. In some MCE 2.4 environments,
+// the managedclusterset and managedclustersetbinding CRDs still have both v1beta1 and v1beta2 versions
+// in etcd and the privous migrations are failed of processing the CRDs.
+//
+// If we upgrade directly from MCE 2.4 to 2.5, the operator will apply the New CRDs(v1beta1 removed)
+// and that will cause crash because the old version CRs still exist in etcd.
+//
+// So before the operator starts, we need to recreate failed migrations to retrigger the migration job.
+// After that, the operator can start safely and apply New CRDs(v1beta1 removed).
+//
+// MigrationPreCheck run before operator controllers. It works as the following:
+//  1. Create the storageversionmigration client.
+//  2. Get ocm-related storageversionmigration CRs. And check the condition of each CR.
+//     If the condition is failed, delete it and create a new one.
+//  3. If all CRs are succeeded, return nil.
+func MigrationPreCheck(ctx context.Context, config *rest.Config) error {
+	klog.Info("Starting Migration Pre Check.")
+	var err error
+
+	// Check if storageversionmigration CRD exists.
+	// If not, return nil.
+	apiExtensionClient, err := apiextensionsclient.NewForConfig(config)
+	if err != nil {
+		return err
+	}
+	_, err = apiExtensionClient.ApiextensionsV1().CustomResourceDefinitions().Get(ctx, "storageversionmigrations.migration.k8s.io", metav1.GetOptions{})
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			return nil
+		}
+		return err
+	}
+	klog.Info("StorageVersionMigration CRD exists.")
+
+	migrationClient, err := migrationv1alpha1client.NewForConfig(config)
+	if err != nil {
+		return err
+	}
+
+check:
+
+	// Get all storageversionmigration CRs.
+	migrations, err := migrationClient.StorageVersionMigrations().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	// Filter the migrations which need to be checked.
+	migrationsNeedCheck := filterMigrations(migrations.Items, checkList)
+
+	// If no migrations need to be checked, return nil.
+	if len(migrationsNeedCheck) == 0 {
+		return nil
+	}
+
+	// Check each migration in the checkList.
+	for _, migration := range migrationsNeedCheck {
+		// If migration succeed, skip.
+		checkResult := checkMigrationStatus(migration)
+		if checkResult {
+			continue
+		}
+
+		// If migration failed, delete the failed migration.
+		err = migrationClient.StorageVersionMigrations().Delete(ctx, migration.Name, metav1.DeleteOptions{})
+		if err != nil {
+			return err
+		}
+
+		// Create a new migration instane to re-do the migration job.
+		migrationNew := migration.DeepCopy()
+		migrationNew.ResourceVersion = ""
+		migrationNew.Status.Conditions = []migrationv1alpha1.MigrationCondition{}
+		_, err = migrationClient.StorageVersionMigrations().Create(ctx, migrationNew, metav1.CreateOptions{})
+		if err != nil {
+			return err
+		}
+
+		goto check
```

Comment:
Use of `goto` statement is discouraged because it makes the program difficult to understand and hard to modify.